app/models/ems_refresh/save_inventory.rb

```diff
@@ -181,6 +181,29 @@ def save_vms_inventory(ems, hashes, target = nil)
     end
   end
 
+  # Convert all mapped hashes into actual tags and associate them with the object.
+  # The hash[:tags] object should be an array of hashes, probably created by
+  # the ContainerLabelTagMapping.map_labels method. Each hash in the array should
+  # have the following basic structure:
+  #
+  # {:category_tag_id=>139, :entry_name=>"foo", :entry_description=>"bar"}
+  #
+  # For now, the object must be a VmOrTemplate object.
+  #
+  def save_tags_inventory(object, hash)
+    return if hash.blank?
+    return unless object.kind_of?(VmOrTemplate)
+
+    tags = hash[:tags].collect do |tag_hash|
+      ContainerLabelTagMapping.find_or_create_tag(tag_hash)
```

Comment:
n+1 — hehe, you found where I buried the bodies ;-\

Confusingly, all SaveInventoryFoo modules are mixed into one namespace, doesn't matter in which file they are.  Please move save_tags_inventory here from save_inventory_container.rb, rather than creating 2nd identically named method.